### PR TITLE
Macros for OSACA (and other *CA) analysis of the latency of specific paths of functions

### PR DESCRIPTION
**Pull request by eggrobin**
Example for Cos(0.1):

Other analysers (uiCA, IACA 3.0 and 2.3, LLVM-MCA) on uops.info, Skylake:
* [loop carried through register](https://uica.uops.info/?code=0fb615000000000fb60d00000000440f282500000000440f282d00000000440f28350000000066450f2fd8760b66450f2fc17604b001eb0233c08844242088542420c644242000410f28c80f54c8410f28c00f28d8410f54dcf20f580d00000000410f54cd66480f7ec848c1e005f20f100c300f57cbf20f104430080f28d00f57d3f2440f5cc1f20f106c30100f28c50f28c8410f28c00f57fff20f10fa0f28f7c4e2f9adf10f28cef20f5ccd0f28e7c4e2f9afe1410f28d00f57c0f20f58d0f2410f59d00f28daf2410f59d80f28c20f28c80f28442440c4e2f1a9442450f20f59eaf20f59e80f28442460c4c2f1a9c7c6442420010f57c9f20f10c80f28c30f28d8c4e2f1a95c2430c6442420010f28c50f28d00f28c3c4e2f9adfaf20f58fc440f28c7f2440f58c6410f28c00f28d0410f54d6410f28c0f20f5cc6f20f5cf8410f28c40f55c7660ffbc2f20f102d00000000660f2fe8760b66410f2fc27604b001eb0233c08844242066450f2ec07b04b001eb0233c088442420c64424200084c9f20f1005000000000f8495fefffff2440f11442428&syntax=hex&uArchs=SKL&tools=uiCA&tools=IACA3&tools=IACA23&tools=llvm&alignment=0&uiCAHtmlOptions=dependencies)
* [loop carried through memory](https://uica.uops.info/?code=f2440f104c24280fb615000000000fb60d00000000440f282d00000000440f283500000000440f283d0000000066450f2fe1760b66450f2fca7604b001eb0233c08844242088542420c644242000410f28d10f54d1410f28c10f28d8410f54ddf20f581500000000410f28c60f54c266480f7ec048c1e005f20f100c300f57cbf20f104430080f28d00f57d3f2440f5cc9f20f106c30100f28c50f28c8410f28c1450f57c0f2440f10c2410f28f8c4e2f9adf90f28cff20f5ccd410f28f0c4e2f9aff1410f28e10f57c0f20f58e0f2410f59e10f28d4f2410f59d10f28c40f28c80f28442440c4e2f1a9442450f20f59e5f20f59e00f284424600f286c2470c4e2f1a9c5c6442420010f57c9f20f10c80f28c20f28d8c4e2f1a95c2430c6442420010f28c40f28d00f28c3c462f9adc2f2440f58c6450f28c8f2440f58cf410f28c10f28d0410f54d7410f28c1f20f5cc7f2440f5cc0410f28c5410f55c0660ffbc2f2440f10050000000066440f2fc0760b66410f2fc37604b001eb0233c08844242066450f2ec97b04b001eb0233c088442420c64424200084c9f20f100d000000000f8484fefffff2440f114c2428&syntax=hex&uArchs=SKL&tools=uiCA&tools=IACA3&tools=IACA23&tools=llvm&alignment=0&uiCAHtmlOptions=dependencies)

OSACA, Cascade Lake loop carried through memory:

Cropped dependency graph:
osaca_dg-cropped

Output:
```
Open Source Architecture Code Analyzer (OSACA) - 0.6.1
Analyzed file:      ..\Principia\functions\Release\x64\sin_cos_test.asm
Architecture:       CSX
Timestamp:          2024-12-31 00:51:00


 P - Throughput of LOAD operation can be hidden behind a past or future STORE instruction
 * - Instruction micro-ops not bound to a port
 X - No throughput/latency information for this instruction in data file


Combined Analysis Report
------------------------
                                       Port pressure in cycles
     |  0   - 0DV  |   1   |  2   -  2D  |  3   -  3D  |  4   |   5   |  6   |  7   ||  CP  | LCD  |
----------------------------------------------------------------------------------------------------
4410 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||  4.0 |  4.0 |   movsd xmm9, QWORD PTR OSACA_input$27[rsp]
4411 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movzx edx, BYTE PTR ?UseHardwareFMA@internal@_fma@numerics@principia@@3_NB ; principia::numerics::_fma::internal::UseHardwareFMA
4412 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movzx ecx, BYTE PTR ?OSACA_loop_terminator@@3_NA
4413 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movaps xmm13, XMMWORD PTR ?sign_bit@masks@internal@_sin_cos@numerics@principia@@3U__m128d@@B
4414 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movaps xmm14, XMMWORD PTR ?mantissa_index_bits@masks@internal@_sin_cos@numerics@principia@@3U__m128d@@B
4415 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movaps xmm15, XMMWORD PTR ?exponent_bits@masks@internal@_sin_cos@numerics@principia@@3U__m128d@@B
4416 |             |       |             |             |      |       |      |      ||      |      |   $OSACA_loop$215:
4417 |             |       |             |             |      |       |      |      ||      |      |   ; Line 196
4418 | 1.00        |       |             |             |      |       |      |      ||      |      |   comisd xmm12, xmm9
4419 |             |       |             |             |      |       |      |      ||      |      | * jbe SHORT $LN19@TestBody
4420 | 1.00        |       |             |             |      |       |      |      ||      |      |   comisd xmm9, xmm10
4421 |             |       |             |             |      |       |      |      ||      |      | * jbe SHORT $LN19@TestBody
4422 | 0.00        | 0.000 |             |             |      | 0.000 | 1.00 |      ||      |      |   mov al, 1
4423 | 0.00        |       | 0.50   0.50 | 0.50   0.50 |      |       | 1.00 |      ||      |      |   jmp SHORT $LN20@TestBody
4424 |             |       |             |             |      |       |      |      ||      |      |   $LN19@TestBody:
4425 | 0.00        | 0.000 |             |             |      | 0.000 | 1.00 |      ||      |      |   xor eax, eax
4426 |             |       |             |             |      |       |      |      ||      |      |   $LN20@TestBody:
4427 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$21[rsp], al
4428 |             |       |             |             |      |       |      |      ||      |      |   ; Line 371
4429 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$25[rsp], dl
4430 |             |       |             |             |      |       |      |      ||      |      |   ; Line 372
4431 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$24[rsp], 0
4432 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\include\cstdlib
4433 |             |       |             |             |      |       |      |      ||      |      |   ; Line 23
4434 |             |       |             |             |      |       |      |      ||  0.0 |  0.0 | * movaps xmm2, xmm9
4435 | 0.33        | 0.333 |             |             |      | 0.333 |      |      ||  1.0 |  1.0 |   andps xmm2, xmm1
4436 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4437 |             |       |             |             |      |       |      |      ||      |      |   ; Line 293
4438 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm9
4439 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm3, xmm0
4440 | 0.33        | 0.333 |             |             |      | 0.333 |      |      ||      |      |   andps xmm3, xmm13
4441 |             |       |             |             |      |       |      |      ||      |      |   ; Line 138
4442 | 0.33        | 0.333 | 0.50   0.50 | 0.50   0.50 |      | 0.333 |      |      ||  4.0 |  4.0 |   addsd xmm2, QWORD PTR __real@42a0000000000000
4443 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm14
4444 | 0.67        | 0.163 |             |             |      | 0.163 |      |      ||  1.0 |  1.0 |   andps xmm0, xmm2
4445 | 1.00        |       |             |             |      |       |      |      ||  1.0 |  1.0 |   movq rax, xmm0
4446 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\include\array
4447 |             |       |             |             |      |       |      |      ||      |      |   ; Line 545
4448 | 0.00        |       |             |             |      |       | 1.00 |      ||  1.0 |  1.0 |   shl rax, 5
4449 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4450 |             |       |             |             |      |       |      |      ||      |      |   ; Line 296
4451 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||  4.0 |  4.0 |   movsd xmm1, QWORD PTR [rax+rsi]
4452 | 0.33        | 0.333 |             |             |      | 0.333 |      |      ||  1.0 |  1.0 |   xorps xmm1, xmm3
4453 |             |       |             |             |      |       |      |      ||      |      |   ; Line 298
4454 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movsd xmm0, QWORD PTR [rax+rsi+8]
4455 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm2, xmm0
4456 | 0.33        | 0.333 |             |             |      | 0.333 |      |      ||      |      |   xorps xmm2, xmm3
4457 |             |       |             |             |      |       |      |      ||      |      |   ; Line 306
4458 | 0.33        | 0.333 |             |             |      | 0.333 |      |      ||  4.0 |  4.0 |   subsd xmm9, xmm1
4459 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\elementary_functions_body.hpp
4460 |             |       |             |             |      |       |      |      ||      |      |   ; Line 60
4461 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movsd xmm5, QWORD PTR [rax+rsi+16]
4462 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\fma_body.hpp
4463 |             |       |             |             |      |       |      |      ||      |      |   ; Line 35
4464 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm5
4465 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm1, xmm0
4466 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm9
4467 | 0.23        | 0.243 |             |             |      | 0.523 |      |      ||      |      |   xorps xmm8, xmm8
4468 |             |       |             |             |      | 1.000 |      |      ||      |      |   movsd xmm8, xmm2
4469 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm7, xmm8
4470 | 0.50        | 0.500 |             |             |      |       |      |      ||      |      |   vfnmadd213sd xmm7, xmm0, xmm1
4471 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\double_precision_body.hpp
4472 |             |       |             |             |      |       |      |      ||      |      |   ; Line 310
4473 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm1, xmm7
4474 | 0.00        | 0.000 |             |             |      | 1.000 |      |      ||      |      |   subsd xmm1, xmm5
4475 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\fma_body.hpp
4476 |             |       |             |             |      |       |      |      ||      |      |   ; Line 46
4477 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm6, xmm8
4478 | 0.50        | 0.500 |             |             |      |       |      |      ||      |      |   vfnmsub213sd xmm6, xmm0, xmm1
4479 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4480 |             |       |             |             |      |       |      |      ||      |      |   ; Line 310
4481 |             |       |             |             |      |       |      |      ||  0.0 |  0.0 | * movaps xmm4, xmm9
4482 | 0.00        | 0.000 |             |             |      | 1.000 |      |      ||      |      |   xorps xmm0, xmm0
4483 | 0.50        | 0.500 |             |             |      |       |      |      ||  4.0 |  4.0 |   addsd xmm4, xmm0
4484 | 0.50        | 0.500 |             |             |      |       |      |      ||  4.0 |  4.0 |   mulsd xmm4, xmm9
4485 |             |       |             |             |      |       |      |      ||      |      |   ; Line 311
4486 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm2, xmm4
4487 | 0.50        | 0.500 |             |             |      |       |      |      ||      |      |   mulsd xmm2, xmm9
4488 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\fma_body.hpp
4489 |             |       |             |             |      |       |      |      ||      |      |   ; Line 14
4490 |             |       |             |             |      |       |      |      ||  0.0 |  0.0 | * movaps xmm0, xmm4
4491 |             |       |             |             |      |       |      |      ||  0.0 |  0.0 | * movaps xmm1, xmm0
4492 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movaps xmm0, XMMWORD PTR tv1368[rsp]
4493 | 0.50        | 0.500 | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   vfmadd213sd xmm0, xmm1, XMMWORD PTR tv1360[rsp]
4494 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4495 |             |       |             |             |      |       |      |      ||      |      |   ; Line 312
4496 | 0.21        | 0.790 |             |             |      |       |      |      ||      |      |   mulsd xmm4, xmm5
4497 | 0.00        | 1.000 |             |             |      |       |      |      ||      |      |   mulsd xmm4, xmm0
4498 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\fma_body.hpp
4499 |             |       |             |             |      |       |      |      ||      |      |   ; Line 14
4500 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movaps xmm0, XMMWORD PTR tv1367[rsp]
4501 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movaps xmm5, XMMWORD PTR tv1356[rsp]
4502 | 0.00        | 1.000 |             |             |      |       |      |      ||  4.0 |  4.0 |   vfmadd213sd xmm0, xmm1, xmm5
4503 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4504 |             |       |             |             |      |       |      |      ||      |      |   ; Line 146
4505 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$20[rsp], 1
4506 | 0.00        | 0.000 |             |             |      | 1.000 |      |      ||      |      |   xorps xmm1, xmm1
4507 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\fma_body.hpp
4508 |             |       |             |             |      |       |      |      ||      |      |   ; Line 14
4509 |             |       |             |             |      | 1.000 |      |      ||  1.0 |  1.0 |   movsd xmm1, xmm0
4510 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm2
4511 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm3, xmm0
4512 | 0.00        | 1.000 | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||  4.0 |  4.0 |   vfmadd213sd xmm3, xmm1, XMMWORD PTR tv1353[rsp]
4513 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4514 |             |       |             |             |      |       |      |      ||      |      |   ; Line 157
4515 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$19[rsp], 1
4516 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\fma_body.hpp
4517 |             |       |             |             |      |       |      |      ||      |      |   ; Line 35
4518 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm4
4519 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm2, xmm0
4520 |             |       |             |             |      |       |      |      ||  0.0 |  0.0 | * movaps xmm0, xmm3
4521 | 0.00        | 1.000 |             |             |      |       |      |      ||  4.0 |  4.0 |   vfnmadd213sd xmm8, xmm0, xmm2
4522 | 0.00        | 1.000 |             |             |      |       |      |      ||  4.0 |  4.0 |   addsd xmm8, xmm6
4523 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\double_precision_body.hpp
4524 |             |       |             |             |      |       |      |      ||      |      |   ; Line 352
4525 |             |       |             |             |      |       |      |      ||  0.0 |  0.0 | * movaps xmm9, xmm8
4526 | 0.00        | 1.000 |             |             |      |       |      |      ||  4.0 |  4.0 |   addsd xmm9, xmm7
4527 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4528 |             |       |             |             |      |       |      |      ||      |      |   ; Line 173
4529 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm9
4530 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm2, xmm0
4531 | 0.22        | 0.207 |             |             |      | 0.573 |      |      ||      |      |   andps xmm2, xmm15
4532 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\double_precision_body.hpp
4533 |             |       |             |             |      |       |      |      ||      |      |   ; Line 353
4534 |             |       |             |             |      |       |      |      ||  0.0 |      | * movaps xmm0, xmm9
4535 | 0.00        | -0.01 |             |             |      | 1.000 |      |      ||  4.0 |      |   subsd xmm0, xmm7
4536 | 0.00        | -0.01 |             |             |      | 1.000 |      |      ||  4.0 |      |   subsd xmm8, xmm0
4537 |             |       |             |             |      |       |      |      ||      |      |   ; File C:\Users\robin\Projects\mockingbirdnest\Principia\numerics\sin_cos.cpp
4538 |             |       |             |             |      |       |      |      ||      |      |   ; Line 175
4539 |             |       |             |             |      |       |      |      ||      |      | * movaps xmm0, xmm13
4540 | 0.00        | -0.01 |             |             |      | 1.000 |      |      ||  1.0 |      |   andnps xmm0, xmm8
4541 |             |       |             |             |      |       |      |      ||      |      |   ; Line 177
4542 | 0.00        | -0.01 |             |             |      | 1.000 |      |      ||  1.0 |      |   psubq xmm0, xmm2
4543 |             |       |             |             |      |       |      |      ||      |      |   ; Line 181
4544 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movsd xmm8, QWORD PTR __real@fcaffff000000000
4545 | 1.00        |       |             |             |      |       |      |      ||  3.0 |      |   comisd xmm8, xmm0
4546 |             |       |             |             |      |       |      |      ||      |      | * jbe SHORT $LN125@TestBody
4547 | 1.00        |       |             |             |      |       |      |      ||      |      |   comisd xmm0, xmm11
4548 |             |       |             |             |      |       |      |      ||      |      | * jbe SHORT $LN125@TestBody
4549 | 0.00        | -0.01 |             |             |      | 0.090 | 0.91 |      ||      |      |   mov al, 1
4550 | 0.00        |       | 0.50   0.50 | 0.50   0.50 |      |       | 1.00 |      ||      |      |   jmp SHORT $LN126@TestBody
4551 |             |       |             |             |      |       |      |      ||      |      |   $LN125@TestBody:
4552 | 0.00        | 0.000 |             |             |      | -0.01 | 1.00 |      ||      |      |   xor eax, eax
4553 |             |       |             |             |      |       |      |      ||      |      |   $LN126@TestBody:
4554 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$18[rsp], al
4555 |             |       |             |             |      |       |      |      ||      |      |   ; Line 384
4556 | 1.00        |       |             |             |      |       |      |      ||      |      |   ucomisd xmm9, xmm9
4557 |             |       |             |             |      |       |      |      ||      |      | * jnp SHORT $LN11@TestBody
4558 | 0.00        | -0.01 |             |             |      | 0.000 | 1.00 |      ||      |      |   mov al, 1
4559 | 0.00        |       | 0.50   0.50 | 0.50   0.50 |      |       | 1.00 |      ||      |      |   jmp SHORT $LN12@TestBody
4560 |             |       |             |             |      |       |      |      ||      |      |   $LN11@TestBody:
4561 | 0.00        | 0.000 |             |             |      | -0.01 | 1.00 |      ||      |      |   xor eax, eax
4562 |             |       |             |             |      |       |      |      ||      |      |   $LN12@TestBody:
4563 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$23[rsp], al
4564 |             |       |             |             |      |       |      |      ||      |      |   ; Line 386
4565 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |      |   mov BYTE PTR OSACA_computed_condition$22[rsp], 0
4566 |             |       |             |             |      |       |      |      ||      |      |   ; Line 389
4567 | 0.00        | -0.01 |             |             |      | 0.000 | 1.00 |      ||      |      |   test cl, cl
4568 |             |       | 0.50   0.50 | 0.50   0.50 |      |       |      |      ||      |      |   movsd xmm1, QWORD PTR __xmm@7fffffffffffffff7fffffffffffffff
4569 |             |       |             |             |      |       |      |      ||      |      | * je $OSACA_loop$215
4570 |             |       | 0.00        | 0.00        | 1.00 |       |      | 1.00 ||      |  0.0 |   movsd QWORD PTR OSACA_result$26[rsp], xmm9

       12.3          12.33   10.0   10.0   10.0   10.0   9.00   12.33   10.9   9.00    63.0   50.0
```

### CODE REVIEW

File: numerics/sin_cos.cpp [hunk 1]
```diff
@@ -14,10 +14,78 @@
...
+
+#define OSACA_RETURN(result)                       \
+  OSACA_loop_carry = (result);                     \
+  if (!OSACA_loop_terminator) {                    \
+    goto OSACA_loop;                               \
+  }                                                \
+  double volatile OSACA_result = OSACA_loop_carry; \
+  IACA_VC64_END;                                   \
+  return OSACA_result
+#define OSACA_IF_(condition)                                           \
```
Comment (pleroy) [inline numerics/sin_cos.cpp:55]: Bad `\` alignment.  Also, let's drop the final `_`.

File: numerics/sin_cos.cpp [hunk 2]
```diff
@@ -110,7 +178,8 @@ inline double DetectDangerousRounding(double const x, double const Δx) {
       _mm_castsi128_pd(_mm_sub_epi64(error_128i, value_exponent_128i)));
   // TODO(phl): Error analysis to refine the thresholds.  Also, can we avoid
   // negative numbers?
-  if (normalized_error < -0x1.ffffp971 && normalized_error > -0x1.00008p972) {
+  OSACA_IF_(normalized_error < -0x1.ffffp971 &&
+          normalized_error > -0x1.00008p972) {
```
Comment (pleroy) [inline numerics/sin_cos.cpp:182]: Bad indent.

File: numerics/sin_cos.cpp [hunk 3]
```diff
@@ -254,71 +323,67 @@ Value CosImplementation(DoublePrecision<Argument> const θ_reduced) {
...
   Reduce(θ, θ_reduced, quadrant);
   double value;
-  if (UseHardwareFMA) {
-    if (quadrant & 0b1) {
+  OSACA_IF_(UseHardwareFMA) {
+    OSACA_IF_(quadrant & 0b1) {
       value = CosImplementation<FMAPolicy::Force>(θ_reduced);
     } else {
 #if PRINCIPIA_USE_OSACA_SIN
-      IACA_VC64_START;
```
Comment (pleroy) [inline numerics/sin_cos.cpp:266]: Remove.

File: numerics/sin_cos.cpp [hunk 4]
```diff
@@ -254,71 +323,67 @@ Value CosImplementation(DoublePrecision<Argument> const θ_reduced) {
...
-    return value;
+    OSACA_RETURN_SIN(value);
   }
 }
 
 #if PRINCIPIA_INLINE_SIN_COS
 FORCE_INLINE(inline)
 #endif
-Value __cdecl Cos(Argument const θ) {
+Value __cdecl Cos(Argument θ) {
```
Comment (pleroy) [inline numerics/sin_cos.cpp:362]: Can we keep the `const`?

Comment (eggrobin) [reply numerics/sin_cos.cpp:358]: No.

File: numerics/sin_cos.cpp [hunk 5]
```diff
@@ -14,10 +14,199 @@
...
+#define PRINCIPIA_USE_OSACA PRINCIPIA_USE_OSACA_SIN || PRINCIPIA_USE_OSACA_COS
+
+#if PRINCIPIA_USE_OSACA
+
 #include "intel/iacaMarks.h"
+
+// The macros OSACA_FUNCTION_BEGIN and OSACA_RETURN are used to analyse the
+// latency of a double -> double function, as measured, e.g., by the
+// nanobenchmarks; this notionally corresponds to the duration of an iteration
+// of a loop x = f(x).
```
Comment (pleroy) [inline numerics/sin_cos.cpp:26]: Backticks.

File: numerics/sin_cos.cpp [hunk 6]
```diff
@@ -14,10 +14,199 @@
...
+    return (expression);                                                   \
+  }()
+#endif
+// In principle, the loop-carried dependency for function latency analysis is
+// achieved by wrapping the body of the function in an infinite loop, assigning
+// the result to the argument at each iteration, and adding IACA markers outside
+// the loop.  There are some subtleties:
+// — We need to trick the compiler into believing the loop is finite, so that it
+//   doesn’t optimize away the end marker or even the function.  This is
+//   achieved by exiting based on the value of OSACA_loop_terminator.
```
Comment (pleroy) [inline numerics/sin_cos.cpp:96]: Backticks.

File: numerics/sin_cos.cpp [hunk 7]
```diff
@@ -14,10 +14,199 @@
...
+// — Return statements may be in if statements, and there may be several of
+//   them, so they cannot be the end of a loop started unconditionally.  Instead
+//   we loop with goto.
+// — Some volatile reads and writes are used to clarify identity of the
+//   registers in the generated code (where the names of OSACA_input and
```
Comment (pleroy) [inline numerics/sin_cos.cpp:101]: More backticks here and in the next line.

File: numerics/sin_cos.cpp [hunk 8]
```diff
@@ -14,10 +14,199 @@
...
+#else
+#define OSACA_INPUT_QUALIFIER volatile
+#endif
+
+// The branch not taken, determined by evaluating the condition
+// `UNDER_OSACA_HYPOTHESES`, is eliminated by `if constexpr`; the condition is
+// also compiled normally and assigned to a boolean; whether this results in any
+// generated code depends on `OSACA_EVALUATE_CONDITIONS`.  Note that, with
+// `OSACA_EVALUATE_CONDITIONS`, in  `OSACA_IF(p) { } OSACA_ELSE_IF(q) { }`, if
+// `p` holds `UNDER_OSACA_HYPOTHESES`, code is generated to evalutae `p`, but
```
Comment (pleroy) [inline numerics/sin_cos.cpp:144]: Typo.

File: numerics/sin_cos.cpp [hunk 9]
```diff
@@ -14,10 +14,199 @@
...
+                    (condition);                                          \
+                UNDER_OSACA_HYPOTHESES(condition))
+
+#if OSACA_EVALUATE_CONDITIONS
+#define OSACA_CONDITION_QUALIFIER volatile
+#else
+#define OSACA_CONDITION_QUALIFIER
+#endif
+
+#else
```
Comment (pleroy) [inline numerics/sin_cos.cpp:158]: Comment the condition of which this is the `else`.

**Pull request merged**